x/cdp/keeper/seize.go

```diff
@@ -93,7 +115,49 @@ func (k Keeper) ApplyLiquidationPenalty(ctx sdk.Context, collateralType string,
 	return sdk.NewDecFromInt(debt).Mul(penalty).RoundInt()
 }
 
+// ValidateLiquidation validate that adding the input principal puts the cdp below the liquidation ratio
+func (k Keeper) ValidateLiquidation(ctx sdk.Context, collateral sdk.Coin, collateralType string, principal sdk.Coin, fees sdk.Coin) error {
+	collateralizationRatio, err := k.CalculateCollateralizationRatio(ctx, collateral, collateralType, principal, fees, spot)
+	if err != nil {
+		return err
+	}
+	liquidationRatio := k.getLiquidationRatio(ctx, collateralType)
+	if collateralizationRatio.GT(liquidationRatio) {
+		return sdkerrors.Wrapf(types.ErrNotLiquidatable, "collateral %s, collateral ratio %s, liquidation ratio %s", collateral.Denom, collateralizationRatio, liquidationRatio)
+	}
+	return nil
+}
+
 func (k Keeper) getModAccountDebt(ctx sdk.Context, accountName string) sdk.Int {
 	macc := k.supplyKeeper.GetModuleAccount(ctx, accountName)
 	return macc.GetCoins().AmountOf(k.GetDebtDenom(ctx))
 }
+
+func (k Keeper) payoutKeeperLiquidationReward(ctx sdk.Context, keeper sdk.AccAddress, cdp types.CDP) (types.CDP, error) {
+	collateralParam, found := k.GetCollateral(ctx, cdp.Type)
+	if !found {
+		return types.CDP{}, sdkerrors.Wrapf(types.ErrInvalidCollateral, "%s", cdp.Type)
+	}
+	reward := cdp.Collateral.Amount.ToDec().Mul(collateralParam.KeeperRewardPercentage).RoundInt()
+	rewardCoin := sdk.NewCoin(cdp.Collateral.Denom, reward)
+	err := k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, keeper, sdk.NewCoins(rewardCoin))
+	if err != nil {
+		return types.CDP{}, err
+	}
+	cdp.Collateral = cdp.Collateral.Sub(rewardCoin)
+	ratio := k.CalculateCollateralToDebtRatio(ctx, cdp.Collateral, cdp.Type, cdp.GetTotalPrincipal())
+	err = k.UpdateCdpAndCollateralRatioIndex(ctx, cdp, ratio)
+	if err != nil {
+		return types.CDP{}, err
+	}
+
+	deposits := k.GetDeposits(ctx, cdp.ID)
+	for _, dep := range deposits {
+		if dep.Amount.IsGTE(rewardCoin) {
+			dep.Amount = dep.Amount.Sub(rewardCoin)
+			k.SetDeposit(ctx, dep)
+			break
+		}
+	}
```

Comment:
If no deposit is greater than the `rewardCoin` then the `if` is never triggered and the reward is not subtracted.

It would leave cdp.Collateral not equal to the sum of deposits. In SeizeCollateral, auctions are started from deposits, rather than cdp.Collateral, so deposits would be auctioned without the reward subtracted. End result is too much collateral is sent out from the cdp module account.

Seems really unlikely to occur naturally, but could be forced by a user.